### DISCUSSION TITLE
[Feature]: Support mTLS / Client TLS Certificates in Mobile App

### DISCUSSION THREAD

**Discussion by JabobKrauskopf** [immich-app/immich#1611] [Feature Request]
### Feature detail

**Background**
I would like to access my instance of Immich from outside my network. However, the only option I have without exposing it to the internet is to use a VPN, which I would like to avoid. Configuring my reverse-proxy to require [mutual TLS / client certificates ](https://www.cloudflare.com/en-gb/learning/access-management/what-is-mutual-tls/) could reduce the risk of exposing Immich by adding an extra layer of protection.

**Proposal**
Ideally, the Immich app could access the iOS / Android keychain and use a client certificate that is installed system-wide for any requests to the Immich server.

**Problem**
After taking a stab at implementing this feature on Android by accessing the native keychain, I ran into the following issue. The Dart HTTP Client supports client certificates by configuring a `SecurityContext` object with `useCertificateChainBytes()` and `usePrivateKeyBytes()`. However, these methods require the actual contents of the private key, but the Android `KeyChain.getPrivateKey()` API only returns an opaque key. The built-in Dart HTTP client does not seem to support this (see https://github.com/dart-lang/sdk/issues/50669#issuecomment-1350624999).

Apps like HomeAssistant support client certificates because the OkHttp library seems to use native Android APIs (see https://github.com/home-assistant/android/pull/2526).

**Solutions**
One solution would be to have some kind of advanced options section on the login screen where a user uploads a client certificate. However, this would be less than ideal since we would bypass the system keystore which already provides this functionality.

Another option that might be worth exploring is a custom header that contains a secret token which is sent with every request. Maybe something like this: https://github.com/immich-app/immich/issues/1305

I would appreciate any insights/help on how one might approach implementing mutual TLS.

### Platform

Mobile App

**Comment by stokito** [comment]
Why not just authorize by username and password? Then you can remember or use a password manager. A long password gives just enough of protection.

The mutual TLS in practice never worked. It looks similar to ssh keys but lacks of all infrastructure  and tools. There is no any reason to use it and that's why it's not supported anywhere.

**Reply by JabobKrauskopf** [reply]
> Why not just authorize by username and password? Then you can remember or use a password manager. A long password gives just enough of protection.

It's not that I don't trust authentication using username and password. But logic bugs can and do exist (not necessarily in Immich but in general) and it is in the realm of possibilities that one might be able to access restricted content using some sort of loop hole or vulnerability in the code.
This matters to me because I want to ensure that my photo collection remains private, due to its sensitive nature.
Being able to secure my Immich instance on the transport layer instead of hoping that there will never be any vulnerabilities in the application layer would be ideal.
Here's an example that these vulnerabilities can exist, even in an established project like Home Assistant: https://github.com/home-assistant/core/pull/13836

> It looks similar to ssh keys but lacks of all infrastructure and tools.

Client certificates are created using the same tools as "normal" server certificates.
Regarding infrastructure, common reverse proxies (traefik, caddy, nginx) support requiring and verifying client certificates. 

> The mutual TLS in practice never worked. [...] There is no any reason to use it and that's why it's not supported anywhere.

I disagree. While this might not be a feature everyone uses, there is clearly at least some demand for it, as can be seen by searching for `mTLS` or `client certificate` in the Home Assistant and Nextcloud issues/PRs (see for example https://github.com/home-assistant/android/issues/27 https://github.com/nextcloud/android/issues/603 https://github.com/nextcloud/android/pull/11314).
Even if not implemented using mTLS there is demand for a zero trust solution for Immich by other people as well: https://github.com/immich-app/immich/issues/1305 .

**Reply by stokito** [reply]
If we are talking about loop hole as it was in the HA when logs wasn't protected then this is a bug on app level. It's easy to make such mistakes and in future they may made again, that's true.
You can make an authorization on proxy level e.g. add a basic auth for entire root. The basic auth may work in parallel to cookie based auth.
The mtls has only two advantages:
1. The key is bigger than a password
2. You can reject a hacking attempts earlier on a transport layer.

Both aren't such a big advantage. A long password from three words anyway gives years to brute force. The TLS handshake is anyway slow by itself but still hard to DDoS it.

When I saying that there is no good tools I mean something like ssh-copy-id or just a specified folder where your keys are stored like `~/.ssh/`. A lot of things are just missing and you have only a key file itself.
This is not critical but you can't make the mTLS widely adopted.
I remember in old days browsers had a support of mutual TLS and even one website (WebMoney) used it. But today you won't find anyone who use the mtls.


If you have a private instance then a simplified attack vector looks like:
1. Some bot scanner may send known exploits. Here the best protection is to keep the app always updated and additional measures like VPN, IP whitelist or even the mTLS.
2. Targeted attack when some hacker tries to attack you personally. But why would someone interested in you and spent time?
3. Mass surveillance e.g. passive eavesdropping

From one side even if you keep your photos just open that doesn't mean that someone will find them immediately and stole. But from another side you better not to upload sensitive information anywhere in Internet

**Reply by bestrocker221** [reply]
@stokito I do not agree.
>Why not just authorize by username and password? Then you can remember or use a password manager. A long password gives just enough of protection.

In the remote case there is a bug in the app authentication, a password can become irrelevant. And what if a specific endpoint is not properly/mistakenly protected?

mTLS authentication is amongst the most secure mechanism to ensure private and trusted communication between two parties.
For self hosting, mTLS almost certainty ensure that nobody else can access your https endpoint, even before touching the web application (unless the certificate is stolen, or CVE in web server mTLS logic). That is one thing that makes mTLS awesome in self hosted environments where, even if you expose an endpoint to the Internet, you know exactly who can access it. (and also why it was such a requested feature in HA)

Web applications can contain vulnerabilities at any time, which puts your data at risk.
Even if you have setup SSO/OpenID/Proxy auth, which might greatly improve things, bugs in the idp or in the app itself can still be discovered, which makes the whole thing fall apart again. People might trust google photos that such bugs/vulnerabilities would be caught in time and never affect their data, but when self hosting their own NNNk+ private picture and video library it becomes a bigger concern.

I do see a mTLS feature (just to be able to reuse certificates from the keystore would be enough) for the immich app of great use. I wish I could help contributing but JS/android is not my field. In the meantime with vpn works great

That said, I want to add that the project seems awesome and thanks for the development!

**Reply by piotrszlenk** [reply]
I think you are on the money here. Consuming an external service like Cloudflare adds the missing piece of protection and privacy. You don't need to expose any ports to the public internet and reveal your public IP. Instead you run a reverse tunnels back to cloudflare so it supports dynamic public IPS/CG-NATs, etc.

The only modification that is required is at the Android/IOS app level.

Thanks for all the development so far.

**Reply by P4SQL** [reply]
Hi all

I just started to use immich two weeks ago, which is awesome and the best photo manager by far, and I started to look into security features which includes mTLS via "any" proxy, in my case it's traefik.

> Why not just authorize by username and password? Then you can remember or use a password manager. A long password gives just enough of protection.

Excuse my french, but WTF(udge)! Even with a strong and secure password, you should ALWAYS consider to have a second factor (ALWAYS!). I highly recommend to have a look into mTLS if you are serious about security and keep the bad people away from your self-hosted apps when they are exposed to the internet.

> The mutual TLS in practice never worked. It looks similar to ssh keys but lacks of all infrastructure and tools. There is no any reason to use it and that's why it's not supported anywhere.

I do not agree at all! mTLS is challenging as you need to have a CA, client certificates and you have to understand a lot about it before you start. All tools are free available and you just have to chose your favorite Linux and make yourself familiar with openssl to setup the infrastructure; the same applies to SSH with key auth.

Bitwarden/Vaultwarden is my password manager for many years and mTLS works great! A lot of self-hosted apps still do not support mTLS, the proxy will take care of it, but the app for Android/iOS has to support client certificates in the first place.

**Reply by Finkregh** [reply]
For everyone who would like to try this, I have wrote down what I did to get the certs as well as the config for caddy: https://git.h.oluflorenzen.de/finkregh/scratchpad/wiki/mTLS-with-cfssl-generated-certificates

**Comment by bo0tzz** [comment]
I think it is unlikely that we will add support for this to Immich. A VPN allows for the same level of protections, while working seamlessly without any application changes.

**Reply by maratsh** [reply]
vpn eats battery and requires separate app on phone plus server. mTls on the other hand is totally passive.

Support for mTls is something I would expect from self hosted application, as it's quite common among them to have this feature. Having proper 'tools' to distribute keys does not really matter. As it's intended for personal usage of few people, it is easy to issue and distribute a few certs. If you are thinking about wider audience support for client certificates, there are plenty of tools to issue and distribute certs for organisations.

**Reply by bo0tzz** [reply]
Modern VPNs like Wireguard use essentially no extra battery power

**Reply by macvenez** [reply]
Also it's way easier if you want to share Immich with other people which are not in your LAN to just send them your mTLS key and install (which on most devices is just a one click operation) and avoid explaining VPN to them and make sure they have it enabled when accessing Immich server

**Reply by IMBeniamin** [reply]
> Modern VPNs like Wireguard use essentially no extra battery power

While this is true I can tell you form experience that setting up a client certificate on my parents' devices is FAR easier and far more reliable than setting up a vpn, which they will most likely either uninstall or disable and then wonder why everything isn't working anymore.

I won't talk about the disadvantages of keeping a vpn that routes all internet traffic consistently on, but even if it does split tunnelling you then have to handle routing when connected to your home network, where the vpn is actually no longer needed and  android does not provide an easy way to manage this.

Using vpn that is always on in my experience has just been too cumbersome for the few times it was actually needed.
That's why I am taking a risk by hosting immich publicly behind a reverse proxy and cloudflare, the vpn setup is just too complicated to be useful.

Oh and did I mention that some networks actively block VPNs? That's a really nice one to try to figure out, whereas a certificate just... works, magically

**Reply by github-cli** [reply]
Some networks also actively block client certificates (I block all incoming traffic that cannot be decrypted)

I read through this but hope if this is considered, it will be configurable.
I don’t see me sharing photos to friends and they install client certs to see shared vacation pictures

**Reply by IMBeniamin** [reply]
> Some networks also actively block client certificates (I block all incoming traffic that cannot be decrypted)
> 
> I read through this but hope if this is considered, it will be configurable.
> I don’t see me sharing photos to friends and they install client certs to see shared vacation pictures 

Some networks do indeed block client certificates, but it is safe to assume, much more likely in fact, that such a network would also block vpns.

I believe the publicly shared links should not be protected by mTLS as they are public. (Yes I know this opens the potential for vulnerabilities, but it's far more usable)

The general rule I have in mind is:
If Auth is required it should be protected my mTLS

**Reply by bestrocker221** [reply]
The mTLS would be an optional feature, and enforced by the web server anyway when setup. So it is up to how you configure your proxy to require mTLS only on certain paths or all of them. (example: allow the shared links)
The whole point here is to allow immich to pick a user certificate when it is required by the proxy.

**Reply by github-cli** [reply]
Not true but obviously either exist.
Couldn’t you set this up on your reverse proxy then?
Edit: I guess you can but you are looking for the support for it on the mobile app?

I rely on being able to decrypt it exactly for security reasons. I scan for zero day and other threats on incoming traffic (and outgoing).

**Reply by IMBeniamin** [reply]
There are many solutions to this problem, mTLS on immich is one of them.
The security layers in regards to the web application are pretty easy and are fairly transparent.

Issues arise however when using the mobile apps and native clients, because it is no longer possible for authentication to be completely transparent.

Cloudflare has many solutions to this such as Argo tunnels, access, firewall rules, reverse proxies can be configured to require authentication, etc.

Bo0tzz is right in saying that the only solution that does not require code modifications on the clients is using a VPN.

All other solutions, custom headers, certificates, etc have to also be implemented in the mobile apps.

**Reply by coxtor** [reply]
All we need is an option for the user to enable to send a certificate to the server during initial login, the certificates are usually stored in the devices certificate store once the user has added them. The rest is handled by the server. mTLS would not prevent you of scanning your traffic if you are in control of your network.

**Reply by coxtor** [reply]
Using mTLS you would protect yourself from zero days as only trusted users could access your server in the first place.

**Reply by bestrocker221** [reply]
@coxtor  
>All we need is an option for the user to enable to send a certificate to the server during initial login

Are you referring to a new method of authenticating to Immich via client certificates? that's a whole another story.

The TLS client authentication happens even before touching the Immich application, and is basically only between the end user and the proxy. It is also easier to implement rather than adding another authentication mechanism in the web application, and it would be transparently enforced, for each request, by the underlying infrastructure. 

End-user <-- mTLS --> Proxy <--> Immich

The big plus for Immich mobile application would be to support the use of already imported user certificates (imported in the mobile, not in the app).

That way you can then setup your own CA and then intercept / decrypt according to your preferences

**Reply by github-cli** [reply]
Using client certificates does not prevent attacks, zero trust means you should not blindly trust. The device with the cert can be compromised, you might share it with someone you think you can trust but cannot, etc.
But yes, of course it is much, much more secure than without client cert and no traffic inspection

**Comment by german-aksenov** [comment]
is it possible to make it like 
https://github.com/MarcoEidinger/ClientCertificateSwiftDemo/tree/main

**Comment by badzz** [comment]
@JabobKrauskopf thanks for bringing this, I just started on immich and my all personal hosting is based on client certificate. So no support for this will be a no go. 
I saw in https://github.com/dart-lang/sdk/issues/50669 that somebody did a http native plugin for dart  : https://github.com/badzz/native_http_client
Is this a viable option ?

**Comment by ghost** [comment]
I would like to see mTLS supported in the app too

**Comment by r01k** [comment]
I concur that supporting client certificates would greatly enhance the security options in Immich.

**Comment by piotrszlenk** [comment]
Hey, consider a scenario where you put your server behind cloudflare and enable mtls there.

It limits the surface of  attack for immich as a valid client cert is required to get to immich api.

It's a common use case for home assistant owners.

**Comment by ported-pw** [comment]
It also does seem risky to me to run an in-development thing like this without an additional layer of protection, like what either this (using client certificates) or e.g. custom headers using Cloudflare Access/Zero Trust allows (https://github.com/immich-app/immich/issues/1305).

**Comment by coxtor** [comment]
Why is the Diskussion closed?

**Comment by faryon93** [comment]
Can we please reopen this issue? I think there ist a real benefit to help improve security when the app could provide client certificates to the server.

**Reply by piotrszlenk** [reply]
Hi @alextran1502, I was wondering if we could reopen the discussion we had earlier. I believe it is important for Immich clients to have this feature.

**Comment by Theprim0** [comment]
Looking forward to this feature

**Comment by Finkregh** [comment]
I;d like to sponsor this specifically, is that possible? As I understand, https://immich.app/blog/2024/immich-core-team-goes-fulltime#i-am-currently-supporting-immich-through-github-sponsors-what-will-happen-to-my-donation there is no sponsoring available at the moment?

edit: Librepay it is: https://liberapay.com/alex.tran1502/

**Comment by yjiang-c** [comment]
Nextcloud just added this feature and will be available in the next release. With Nextcloud memories plugin, it can achieve most of Immich functions except for AI. I do not understand why Immich refuse to implement this feature. There are high demand for this feature. If user name and password is safe enough, why do we need 2FA. Adding more security choice is good especially for this intensive developments project.

https://github.com/nextcloud/ios/pull/2908

**Reply by github-cli** [reply]
mhh?
Immich main functions are AI...
and even not considering those, Memories does not come close (or did it recently get a huge update?)

I do not get your reasoning at all, its cynical and does not make sense to me

**Reply by yjiang-c** [reply]
Yes, Immich has AI which is an excellent. But why cannot add client certificate feature. On mobile, installing VPN is quite complicated and lack flexibility. If client cert is not option, do we have 2FA to increase security?

**Reply by github-cli** [reply]
yes, you can use OAuth for 2FA

on a sidenote, I believe installing a VPN on a mobile device is much easier than installing a client certificate for a mobile device!

**Reply by yjiang-c** [reply]
I agree that vpn could easier to be setup than client certificates. But it need a VPN app and keep it running for 24 hours. I am not sure how reliable it will be. In addition, my family member may have no any idea about IT technology, I need to help them to setup VPN and make sure it will be reliable, which seems to more difficult than just install client certificate.

**Reply by github-cli** [reply]
have you ever installed a trusted client certificate on a family members mobile device?
for this use-case, a vpn will surely be easier!

if you are using a client certificate, you are exposing your immich anyways...
a client certificate would protect you from prying eyes and it is much more difficult to decrypt that traffic.

does it actually give your more protection from your immich instance being "hacked" over a very secure password or 2FA?
I do not believe so..

either way, this will be my last response to this.

**Reply by yjiang-c** [reply]
I have ever modified Immich Android client’s source code to enable a client certificate and tested it to work well. But I have no environment to do with iOS client.

**Reply by badzz** [reply]
> have you ever installed a trusted client certificate on a family members mobile device? for this use-case, a vpn will surely be easier!
> 
No, installing a client cert is a piece of cake on family member android phones. I have done it many times. No app to install, no config.. 
I understand the technical difficulties but the requirement obviously exists if you check above.
I really want to use this software (it is really nice!) but don't because of this missing feature.
Edit : sorry for the harsh tone before edit

**Reply by faryon93** [reply]
> if you are using a client certificate, you are exposing your immich anyways...
> a client certificate would protect you from prying eyes and it is much more difficult to decrypt that traffic.

In my scenario a reverse proxy is handling the client certificate. So Immich is not exposed to the public until the client provides a valid certificate. As Immich is a really fast developing project (and we are thankful for that, because everyone likes new features). I assume at some point things break and they may break in a way leaving my personal photos open to the public. I don't say that will happen. I just say that can happen as we have seen authentication bypass bugs in OSS and comercial applications in the past.
Enabling the app to send a client certificate would bei an easy way of adding another layer of security. The odds of the reverse proxy having a critical client certificate bypass bug at the same time as Immich has an authentication bypass bug are pretty low in my opinion.
This feature request is about minimizing the attack surface to the reverse proxy and taking Immich out of the equation.

**Reply by Finkregh** [reply]
>  This feature request is about minimizing the attack surface to the reverse proxy and taking Immich out of the equation.

Exactly this!

**Reply by lpcvoid** [reply]
> have you ever installed a trusted client certificate on a family members mobile device? for this use-case, a vpn will surely be easier!

Installing a TLS cert is literally a one tap operation (on Android at least), and is then safeguarded by the HSM. Adding mTLS auth to the app would be really, really good (and other applications have done the same, as noted).

> if you are using a client certificate, you are exposing your immich anyways... a client certificate would protect you from prying eyes and it is much more difficult to decrypt that traffic.

If my reverse proxy checks client cert, then no, I am not exposing immich if the client isn't validated.

> does it actually give your more protection from your immich instance being "hacked" over a very secure password or 2FA? I do not believe so..
> either way, this will be my last response to this.

Of course - if an endpoint is not reachable, it is not exploitable. How can you argue against this?!

**Reply by coxtor** [reply]
I really don't understand this reasoning either. For my usecase it is simple: In my family / friend domain I am the trusted IT-Guy, I have installed my certificates where they are wanted and have my services exposed to the people that need them. The service is non-existent to other partys so a perfect vpnless solution for all. Please reconsider @github-cli 
Would you be willing to merge a pull request implementing this feature? Maybe somebody who has the skill to do this would be willing to make the contribution @yjiang-c seems to already have made some efforts. Perhaps he could share the code also?

**Reply by github-cli** [reply]
i am just a user like you...
and I misunderstood this here, I figured it would be a check that is done on the server side (immich)
if this is checked by the reverse proxy before having access to immich (how would image sharing work then?), then yes it would definitely provide more security

**Reply by DarkChemicalDC** [reply]
Yes mTLS certificate check is done at the proxy level.
Fun fact, because all of the browsers support mTLS the setup works on all of them. And just the phone app doesn't support that.

**Reply by kylmp** [reply]
> i am just a user like you... and I misunderstood this here, I figured it would be a check that is done on the server side (immich) if this is checked by the reverse proxy before having access to immich (how would image sharing work then?), then yes it would definitely provide more security

Unfortunately mTLS is often confused by many people. The Immich server doesn't need to do anything, the reverse proxy is what requires the certificate before proxying request to Immich.

What we do need however, is for the *client applications* to include this as an option for authentication. The client application needs to use the device's cert store to let the user select which cert they want to send in the requests to the server. Then all future requests the client app sends to Immich will send the cert as well, and that is the means of authentication. iOS and Android should have APIs build into their app development SDKs for this to access the device installed certificates.. 

As for complexity for non technical users the set up is very easy. I just need to put one of my self signed certificates on their device and install it (one tap). 

If I decide I like Immich enough I might try implementing this on the Android app myself, as that is the only client app I would need remote access with. Right now though since I haven't committed to any software yet it might be a deal breaker for the time being personally.

edit: It seems like maybe this feature was implemented for the android app (thanks to @yjiang-c) so I will be checking this out

**Reply by macvenez** [reply]
> > i am just a user like you... and I misunderstood this here, I figured it would be a check that is done on the server side (immich) if this is checked by the reverse proxy before having access to immich (how would image sharing work then?), then yes it would definitely provide more security
> 
> Unfortunately mTLS is often confused by many people. The Immich server doesn't need to do anything, the reverse proxy is what requires the certificate before proxying request to Immich.
> 
> What we do need however, is for the _client applications_ to include this as an option for authentication. The client application needs to use the device's cert store to let the user select which cert they want to send in the requests to the server. Then all future requests the client app sends to Immich will send the cert as well, and that is the means of authentication. iOS and Android should have APIs build into their app development SDKs for this to access the device installed certificates..
> 
> As for complexity for non technical users the set up is very easy. I just need to put one of my self signed certificates on their device and install it (one tap).
> 
> If I decide I like Immich enough I might try implementing this on the Android app myself, as that is the only client app I would need remote access with. Right now though since I haven't committed to any software yet it might be a deal breaker for the time being personally.
> 
> edit: It seems like maybe this feature was implemented for the android app (thanks to @yjiang-c) so I will be checking this out

Yes, that's already been implemented and I have to say it's working flawlessly.
I'm using Traefik for mtls certificate

**Comment by alextran1502** [comment]
I want to let you guys know that this is a welcome feature if you guys can help integrate it into the app. It is just not the priority of the core team to work on at the moment.

FWIW, we just merged customer header support for the mobile app in https://github.com/immich-app/immich/pull/10588

**Reply by faryon93** [reply]
> I want to let you guys know that this is a welcome feature if you guys can help integrate it into the app. It is just not the priority of the core team to work on at the moment.
> 
> FWIW, we just merged customer header support for the mobile app in #10588

Maybe @yjiang-c can share the Android Implementation and we merge them into the official build? I can help with testing :)

**Reply by yjiang-c** [reply]
My change needs to import SSL client key manually to Immich app and cannot use client certificate imported to Android system. 

In ideal case, a SSL client certificate should be imported to Android system once, then different apps like chrome web browser, Immich etc apps can use the imported client certificate from Android system. But Immich uses Flutter programming language, which does not support access to SSL client certificate imported to Android/iOS system.

Do we allow this solution to use copy and past client certificate to Immich App?

**Reply by badzz** [reply]
For the moment, whatever works is ok with me. Can you describe how you did this (it involves recompiling the app?). Once we understand how it could work , we could see if it is viable for everyone.
Thanks

**Reply by yjiang-c** [reply]
The following source code changes will be required and therefore, it involving recompiling the app
1. Add a button in advanced setting to import SSL client certificate
2. Once the button is clicked, it will show a file picker to pick the p12 file including SSL client certificate and private key
3. After the p12 file is selected, prompt a password input to accept p12 file password. If there is no password, just keep it empty.
4. The SSL client certificate bytes will be saved to Immich asset and used in HttpClient.

**Reply by yjiang-c** [reply]
In long term, we should let Immich App to support to use SSL client certificate stored in Android/iOS system. But this need  to replace Flutter http client by Android/iOS native Http Client, which will be a big changes.

**Reply by coxtor** [reply]
It would be great if you could share  your code @yjiang-c

**Reply by yjiang-c** [reply]
I am tidying up my code and will try to create pull request in a few of days to see whether it is acceptable. Once pull request is created, I will post URL here.

**Reply by yjiang-c** [reply]
![Screenshot_1719998977](https://github.com/immich-app/immich/assets/66518086/02f8ece6-193c-466e-bd9e-e8e1692a25af)

**Reply by P4SQL** [reply]
Woooho! This looks interesting.

**Comment by yjiang-c** [comment]
I created a PR https://github.com/immich-app/immich/pull/10860 for this feature. The PR is being review and hopefully can be approved
